include/dlaf/matrix/copy.h

```diff
@@ -77,6 +78,32 @@ void copy(Matrix<const T, Source>& source, Matrix<T, Destination>& dest) {
   copy(source.distribution().localNrTiles(), LocalTileIndex(0, 0), source, LocalTileIndex(0, 0), dest);
 }
 
+namespace internal {
+template <class T, Device Source, Device Destination>
+void copy(MatrixRef<const T, Source>& src, MatrixRef<T, Destination>& dst) {
+  namespace ex = pika::execution::experimental;
+
+  if constexpr (Source == Destination) {
+    if (&src == &dst)
```

Comment:
I don't know if this needs to be resolved in this PR, but since this compares `MatrixRef` addresses and not `Matrix` addresses this could in the worst case lead to deadlocks if the `MatrixRef`s refer to the same `Matrix` (`read` and `readwrite` into the same task). I think one step in the right direction would be to have `operator==` on `MatrixRef` (which compares the addresses of the contained `Matrix` and sub distributions), but I'm not sure if it covers all possible issues.

I'd be fine with dealing with this outside of this PR.

Reply:
That's a good point. I did a no-brainer copy-paste, but actually it hides a dangerous thing. How should we address this? Immediately or just open an issue? @rasolca @msimberg

Reply:
I think that Submatrix pipeline should be considered as well, so it is better to work on it separately.
However it should be taken care as the tridiagonal solver might be impacted by it.
(device <-> host mirror copies in the case of multicore are skipped as they are the same matrix, however after the cleanup they will be two equal ref but different objects)

Reply:
https://github.com/eth-cscs/DLA-Future/issues/1035

Reply:
Thanks @msimberg for opening the issue!

I'll not resolve this comment hoping it will be easier in the future to spot it in case of need of reference.